### PR TITLE
Content cloning and translation

### PR DESCRIPTION
**Pull request by TFleury**
This PR adds new clone handling to replace export/import logic, and also allow translating a clone of the master content item.

It introduces new `Cloning` and `Cloned` content handlers to specify a clone logic that can be different of import/export ones.

`ContentManager`'s `Duplicate` method don't save a new draft content any more, but create a new content initialized by clone handlers. Admin Content Controller Duplicate action opens a editor to edit the cloned content before saving.

It fixes : #5535, #5089, #4983, #4970, #4538

### CODE REVIEW

File: src/Orchard/ContentManagement/IContentManager.cs [hunk 1]
```diff
@@ -42,6 +42,14 @@ public interface IContentManager : IDependency {
         ContentItem Clone(ContentItem contentItem);
 
         /// <summary>
+        /// Makes a clone of the content item with a special context
+        /// </summary>
+        /// <param name="contentItem">The content item to clone</param>
+        /// <param name="cloneContext">The cloning context</param>
+        /// <returns>Clone of the item</returns>
+        ContentItem Clone(ContentItem contentItem, Handlers.CloneContentContext cloneContext);
```
Comment (sebastienros) [inline src/Orchard/ContentManagement/IContentManager.cs]: Why is it necessary?

**Comment by sebastienros** [comment]
It needs to target dev, it's a breaking change, big new feature. But it looks great.

File: src/Orchard/ContentManagement/IContentManager.cs (hunk 1 above)
Comment (TFleury) [inline src/Orchard/ContentManagement/IContentManager.cs]: This implementation is used by Localization. It's needed for Translate (Localization), which uses `Clone` with a `TranslateContentContext` (derived from `CloneContentContext`).

**Comment by Skrypt** [comment]
There is something missing. When a content item has it's culture undefined. If we try to clone the content item it throws an error.

File: src/Orchard/ContentManagement/IContentManager.cs (hunk 1 above)
Comment (sebastienros) [inline src/Orchard/ContentManagement/IContentManager.cs]: From what I see and understand it shouldn't be necessary. Clone doesn't have to know about localization. The localization module should just call Clone(item) then change the new content item accordingly.

File: src/Orchard.Web/Modules/Orchard.Localization/Handlers/TranslateContentContext.cs [hunk 2]
```diff
@@ -0,0 +1,12 @@
+﻿using Orchard.ContentManagement;
+using Orchard.ContentManagement.Handlers;
+using Orchard.Localization.Records;
+
+namespace Orchard.Localization.Handlers {
+    public class TranslateContentContext : CloneContentContext {
+        public CultureRecord TagetCulture { get; set; }
```
Comment (Skrypt) [inline src/Orchard.Web/Modules/Orchard.Localization/Handlers/TranslateContentContext.cs]: TargetCulture "typo"

File: src/Orchard.Web/Modules/Orchard.Localization/Drivers/LocalizationPartDriver.cs [hunk 3]
```diff
@@ -145,8 +146,17 @@ public class LocalizationPartDriver : ContentPartDriver<LocalizationPart> {
...
+        protected override void Cloning(LocalizationPart originalPart, LocalizationPart clonePart, CloneContentContext context) {
+            if(context is TranslateContentContext) {
+                var translateContext = context as TranslateContentContext;
+                if(translateContext.TagetCulture != null) {
+                    clonePart.Culture = translateContext.TagetCulture;
+                }
+                clonePart.MasterContentItem = translateContext.ContentItem;
+            }
+            else {
+                clonePart.Culture = originalPart.Culture;
```
Comment (Skrypt) [inline src/Orchard.Web/Modules/Orchard.Localization/Drivers/LocalizationPartDriver.cs]: if originalPart.Culture != null here

File: src/Orchard/ContentManagement/IContentManager.cs (hunk 1 above)
Comment (Skrypt) [inline src/Orchard/ContentManagement/IContentManager.cs]: He needs to know wich culture can be selected in that new translated "cloned" content item.

**Comment by TFleury** [comment]
As @sebastienros asked, I reopened a new PR based on dev branch. I also removed `TranslateContentContext`.

**Pull request closed without merging**